File: backend/src/main/java/com/medcore/features/billing/service/BillingService.java
package com.medcore.features.billing.service;

import com.medcore.common.response.ApiResponse;
import com.medcore.common.response.PageResponse;
import com.medcore.features.billing.dto.request.AddBillItemRequest;
import com.medcore.features.billing.dto.request.CreateBillRequest;
import com.medcore.features.billing.dto.request.PaymentRequest;
import com.medcore.features.billing.dto.response.BillItemResponse;
import com.medcore.features.billing.dto.response.BillResponse;
import com.medcore.features.billing.dto.response.PaymentResponse;

public interface BillingService {

    ApiResponse<BillResponse> createBill(
            CreateBillRequest request
    );

    ApiResponse<BillItemResponse> addBillItem(
            Long billId,
            AddBillItemRequest request
    );

    ApiResponse<BillResponse> getBillById(
            Long billId
    );

    ApiResponse<BillItemResponse> updateBillItem(
            Long billId,
            Long itemId,
            AddBillItemRequest request
    );

    ApiResponse<Void> deleteBillItem(
            Long billId,
            Long itemId
    );
    
    ApiResponse<PaymentResponse> payBill(
            Long billId,
            PaymentRequest request
    );
    
    ApiResponse<PageResponse<BillResponse>> getHospitalBills(
            int page,
            int size,
            String sortBy,
            String sortDir
    );
    
    ApiResponse<PageResponse<BillResponse>> getOutstandingBills(
            int page,
            int size,
            String sortBy,
            String sortDir
    );
}

File: frontend/src/services/billing.service.js
import { apiFetch } from "@/lib/api-client";

async function handleResponse(response) {
  const result = await response.json();

  if (!response.ok || !result.success) {
    throw new Error(result.message || "Something went wrong");
  }

  return result;
}

export async function getHospitalBills({
  page = 0,
  size = 20,
  sortBy = "billDate",
  sortDir = "desc",
} = {}) {
  const params = new URLSearchParams({
    page: String(page),
    size: String(size),
    sortBy,
    sortDir,
  });
  const response = await apiFetch(`/billing?${params.toString()}`);
  return handleResponse(response);
}

export async function getOutstandingBills({
  page = 0,
  size = 20,
  sortBy = "billDate",
  sortDir = "desc",
} = {}) {
  const params = new URLSearchParams({
    page: String(page),
    size: String(size),
    sortBy,
    sortDir,
  });
  const response = await apiFetch(
    `/billing/outstanding?${params.toString()}`
  );
  return handleResponse(response);
}

export async function getBillById(billId) {
  const response = await apiFetch(`/billing/${billId}`);
  return handleResponse(response);
}

export async function createBill(data) {
  const response = await apiFetch("/billing", {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify(data),
  });
  return handleResponse(response);
}

export async function addBillItem(billId, data) {
  const response = await apiFetch(`/billing/${billId}/items`, {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify(data),
  });
  return handleResponse(response);
}

export async function updateBillItem(billId, itemId, data) {
  const response = await apiFetch(`/billing/${billId}/items/${itemId}`, {
    method: "PUT",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify(data),
  });
  return handleResponse(response);
}

export async function deleteBillItem(billId, itemId) {
  const response = await apiFetch(`/billing/${billId}/items/${itemId}`, {
    method: "DELETE",
  });
  return handleResponse(response);
}

export async function payBill(billId, data) {
  const response = await apiFetch(`/billing/${billId}/payments`, {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify(data),
  });
  return handleResponse(response);
}